### ISSUE TITLE
Nameof should support more constructs

### ISSUE THREAD

**Issue by cartermp** [dotnet/fsharp#7373]
Nameof currently fails to work with three major areas:

* Generic type parameters
* Instance members (without having an instance)
* Overloaded methods

**Comment by dsyme** [comment]
For preview:
* The first two are design for the RFC
* The third can be done via a type annotation to select one overload:

```fsharp

type MethodGroupNameOfTests() =
    member this.MethodGroup() = ()    
    member this.MethodGroup(i:int) = ()

    member this.MethodGroup1(i:int, f:float, s:string) = 0
    member this.MethodGroup1(f:float, l:int64) = "foo"
    member this.MethodGroup1(u:unit -> unit -> int, h: unit) : unit = ()

    member this.``single argument method group name lookup`` () =
        let b = nameof(this.MethodGroup)
        Assert.AreEqual("MethodGroup",b)

    member this.``multiple argument method group name lookup`` () =
        let b = nameof(this.MethodGroup1 : (float * int64 -> _))
        Assert.AreEqual("MethodGroup1",b)
```

**Comment by cartermp** [comment]
Note that for the last one, C# doesn't do anything special. For any overload named `M`, the name is just `"M"`. So I'd expect a similar experience here.

The first two are also accomplishable and I've seen feedback about supporting them already, so it's quite likely that it'd be unexpected that we cannot get a name from those

**Comment by dsyme** [comment]
* Generic type parameters - We decided somewhere along the way that generic type parameters were not included because in `nameof expr` the argument is a syntactic expression (which can't be a type e.g. `'T`), and it wasn't worth the complexity/risk to fiddle with the parser for this case.  If you want to revisit that's fine, though TBH I think just accepting the feature limitation as by-design seems OK.

* Instance members - I'm fine with you guys changing that..  The community implementation of this feature made a simplifying decision early on to reuse the existing logic for name resolution as much as possible which is why the preview requires `cobj.InstanceMember`. 

* Overloaded members - yes a type annotation is currently needed.  Again, the community implementation made a simplifying decision to actually resolve to a specific overloaded member rather than the general set of overloads.  But if you want to change it that's ok by me. (From my perspective I don't expect this is really going to get used much in F#. Also I suggested to the community implementer that we we want to avoid re-implementing the name resolution logic just for this feature)
